ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducer.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Used to handle transactional writes to the command topic
+ */
+public class TransactionalProducer {
+
+  private final TopicPartition commandTopicPartition;
+  private final String commandTopicName;
+
+  private final Consumer<CommandId, Command> commandConsumer;
+  private final Producer<CommandId, Command> commandProducer;
+  private final CommandRunner commandRunner;
+
+  public TransactionalProducer(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Map<String, Object> kafkaConsumerProperties,
+      final Map<String, Object> kafkaProducerProperties
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+            Objects.requireNonNull(commandTopicName, "commandTopicName"),
+            0
+    );
+
+    this.commandConsumer = new KafkaConsumer<>(
+        Objects.requireNonNull(kafkaConsumerProperties, "kafkaConsumerProperties"),
+        InternalTopicJsonSerdeUtil.getJsonDeserializer(CommandId.class, true),
+        InternalTopicJsonSerdeUtil.getJsonDeserializer(Command.class, false)
+    );
+
+    this.commandProducer = new KafkaProducer<>(
+        Objects.requireNonNull(kafkaProducerProperties, "kafkaProducerProperties"),
+        InternalTopicJsonSerdeUtil.getJsonSerializer(true),
+        InternalTopicJsonSerdeUtil.getJsonSerializer(false)
+    );
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+  }
+
+  @VisibleForTesting
+  TransactionalProducer(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+        Objects.requireNonNull(commandTopicName, "commandTopicName"),
+        0
+    );
+    this.commandConsumer = Objects.requireNonNull(commandConsumer, "commandConsumer");
+    this.commandProducer = Objects.requireNonNull(commandProducer, "commandProducer");
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+  }
+
+
+  /** begins transaction */
+  public void begin() {
+    commandConsumer.assign(Collections.singleton(commandTopicPartition));
+    commandProducer.initTransactions();
+    commandProducer.beginTransaction();
+  }
+
+  public void waitForCommandRunner() {
+    final long endOffset = getEndOffset();
+
+    try {
+      int retries = 0;
+      while (commandRunner.getNumCommandProcessed() < endOffset) {
```

Comment:
I don't think we want to use the number of commands processed here. We should just be able to ask the command runner which offset its consumed up to. After all, this is really what we care about and protects us from bugs with bookkeeping the number of commands that we've run.

Reply:
uses offsets now